### PR TITLE
Change message filters to hold a limited number of hashes of messages.

### PR DESCRIPTION
**Pull request by Fraser999**

### CODE REVIEW

File: src/node.rs [hunk 1]
```diff
@@ -285,10 +288,21 @@ impl Node {
...
-        messages.retain(|msg| !self.message_filter.contains(msg));
-        self.message_filter.extend(messages.clone());
-        messages
+    fn filter_messages(&mut self, messages: Vec<Message>) -> Vec<Message> {
+        let mut filtered = vec![];
+        for message in messages {
+            let mut hasher = DefaultHasher::new();
+            message.hash(&mut hasher);
+            let hash = hasher.finish();
+            if !self.message_filter.contains(&hash) {
```
Comment (afck) [inline src/node.rs]: But `VecDeque::contains` is probably O(n)? Maybe the message filter should be an `LruCache` instead.

Comment (michaelsproul) [inline src/node.rs]: We could even try a `HashMap` for `O(1)` average rather than `O(log n)` lookups?

Comment (Fraser999) [inline src/node.rs]: I did consider the `LruCache` (and `LruTimeCache`, `LinkedHashMap`, ring buffers) but I feel that the memory consumption is a bigger problem than the performance.  Perhaps we can discuss further in a hangout.

Comment (michaelsproul) [inline src/node.rs]: Yes, I agree, profiling doesn't show message filtering to be a major drain on cycles, so it's just the mem usage we should aim to reduce...

**Comment by michaelsproul** [comment]
This works great. With block pruning also enabled I was able to run a simulation with 160 nodes in 10 minutes, with peak memory usage of ~2GB. Profiling a run with 100 nodes reveals that filtering messages only uses 0.4% of the run time, so the `O(n = 1024)` indexing on `VecDeque` clearly isn't an issue.

**Pull request merged**